lib/src/front_end/piece_factory.dart

```diff
@@ -219,6 +219,142 @@ mixin PieceFactory {
+                initialization: null,
+                leftSeparator: Token(precedingComments: null),
+                condition: null,
+                rightSeparator: Token(precedingComments: null),
+                updaters: NodeList(isEmpty: true),
+              ) &&
+              var forParts
+          when rightParenthesis.precedingComments == null:
+        return buildPiece((b) {
+          b.token(leftParenthesis);
+          b.token(forParts.leftSeparator);
+          b.token(forParts.rightSeparator);
+          b.token(rightParenthesis);
+        });
+
+      case ForParts forParts &&
+            ForPartsWithDeclarations(variables: AstNode? initializer):
+      case ForParts forParts &&
+            ForPartsWithExpression(initialization: AstNode? initializer):
+        // In a C-style for loop, treat the for loop parts like an argument list
+        // where each clause is a separate argument. This means that when they
+        // split, they split like:
+        //
+        // ```
+        // for (
+        //   initializerClause;
+        //   conditionClause;
+        //   incrementClause
+        // ) {
+        //   body;
+        // }
+        // ```
+        var partsList =
+            DelimitedListBuilder(this, const ListStyle(commas: Commas.none));
+        partsList.leftBracket(leftParenthesis);
+
+        // The initializer clause.
+        if (initializer != null) {
+          partsList.addCommentsBefore(initializer.beginToken);
+          partsList.add(buildPiece((b) {
+            b.visit(initializer);
+            b.token(forParts.leftSeparator);
+          }));
+        } else {
+          // No initializer, so look at the comments before `;`.
+          partsList.addCommentsBefore(forParts.leftSeparator);
+          partsList.add(tokenPiece(forParts.leftSeparator));
+        }
+
+        // The condition clause.
+        if (forParts.condition case var conditionExpression?) {
+          partsList.addCommentsBefore(conditionExpression.beginToken);
+          partsList.add(buildPiece((b) {
+            b.visit(conditionExpression);
+            b.token(forParts.rightSeparator);
+          }));
+        } else {
+          partsList.addCommentsBefore(forParts.rightSeparator);
+          partsList.add(tokenPiece(forParts.rightSeparator));
+        }
+
+        // The update clauses.
+        if (forParts.updaters.isNotEmpty) {
+          partsList.addCommentsBefore(forParts.updaters.first.beginToken);
+          partsList.add(createList(forParts.updaters,
+              style: const ListStyle(commas: Commas.nonTrailing)));
+        }
+
+        partsList.rightBracket(rightParenthesis);
+        return partsList.build();
+
+      case ForPartsWithPattern():
+        throw UnimplementedError();
+
+      case ForEachParts forEachParts &&
+            ForEachPartsWithDeclaration(loopVariable: AstNode variable):
+      case ForEachParts forEachParts &&
+            ForEachPartsWithIdentifier(identifier: AstNode variable):
```

Comment:
[optional] 
```suggestion
      case ForEachParts forEachParts &&
            (ForEachPartsWithDeclaration(loopVariable: AstNode variable) ||
                ForEachPartsWithIdentifier(identifier: AstNode variable)):
```

Reply:
I think I like the slightly flatter original code.